src/main/java/org/testng/Assert.java

```diff
@@ -152,13 +152,13 @@ private static boolean areNotEqualImpl(Object actual, Object expected) {
     if ((expected == null) && (actual == null)) {
       return false;
     }
-    if (expected == null ^ actual == null) {
-      return true;
+    if (expected == null) {
+      return actual.equals(actual);
     }
-    if (!expected.equals(actual) && !actual.equals(expected)) {
-      return true;
+    if (actual == null) {
+      return expected.equals(expected);
     }
-    return false;
+    return !expected.equals(actual);
```

Comment:
The correct implementation was in 2d3ff4efb5283de79a89ef584874340f998dcaee
GitHub does not let me create suggestion comments for deleted lines, therefore here is the raw code for the method body:
```java
    if ((expected == null) && (actual == null)) {
      return false;
    }
    if (expected == null) {
      return !actual.equals(expected);
    }
    return !expected.equals(actual);
```